### PR TITLE
Redis connection: support TLS

### PR DESCRIPTION
**Pull request by Al2Klimov**

### CODE REVIEW

**Comment by Al2Klimov** [comment]
## Test protocol

### Poor man's TLS

Bildschirmfoto 2020-12-11 um 18 39 58

### Server authn

Bildschirmfoto 2020-12-11 um 18 45 34
Bildschirmfoto 2020-12-11 um 18 45 56

### Client authn

Bildschirmfoto 2020-12-11 um 18 48 10

```
4810:M 11 Dec 2020 17:48:03.005 # Error accepting a client connection: error:1417C086:SSL routines:tls_process_client_certificate:certificate verify failed
```

Bildschirmfoto 2020-12-11 um 18 49 35

### "I'm... Howard. Potts."

Bildschirmfoto 2020-12-11 um 18 50 58
Bildschirmfoto 2020-12-11 um 18 51 09

**Comment by lippserd** [comment]
@nilmerg I'm not quite sure why our `PSR2.Methods.MethodDeclaration.Underscore` exclude does not work for PHP CodeSniffer here. Could you have a look please?

**Comment by nilmerg** [comment]
No idea. Maybe because it's not part of this repository's style rules?

But to pose another question here, is it really necessary to utilize Zend framework here again? This would be the first code utilizing it, can't we provide our own implementation?

**Review by lippserd — DISMISSED** [review]
If `Use TLS` is enabled, certificates must be provided, which means we are not allowed to use insecure connections by default.

**Comment by Al2Klimov** [comment]
_By default_ or _at all_?

**Review by nilmerg — CHANGES_REQUESTED** [review]
We shouldn't depend on `Zend_Validate_Abstract`. Implement the validators as part of https://github.com/Icinga/ipl-validator instead. Maybe rewrite the `RedisConfigForm` to use `ipl\Html\Form` instead to avoid making another wrapper.

**Comment by Al2Klimov** [comment]
I consider moving away from Zend Forms in this already existing form another construction area.

**Comment by nilmerg** [comment]
I don't. You're introducing new validators here. We have a dedicated project for that, which doesn't depend on any Zend interface or abstract class. If it's required to move away from `Icinga\Web\Form` to accomplish that, so be it. With the usage of `Zend_Validate_Abstract` I'm not gonna merge this.

**Comment by lippserd** [comment]
> _By default_ or _at all_?

I think we should add a check box that reflects the `--insecure` cURL flag and display that check box once the validation fails due to certificate errors. @Al2Klimov @nilmerg What do you think?

**Comment by nilmerg** [comment]
> > _By default_ or _at all_?
> 
> I think we should add a check box that reflects the `--insecure` cURL flag and display that check box once the validation fails due to certificate errors. @Al2Klimov @nilmerg What do you think?

:+1:

**Comment by nilmerg** [comment]
When can we expect further progress here?

**Comment by Al2Klimov** [comment]
_As part of https://github.com/Icinga/ipl-validator_? I.e. I shall place them _there_?

**Comment by nilmerg** [comment]
> We shouldn't depend on `Zend_Validate_Abstract`. Implement the validators as part of https://github.com/Icinga/ipl-validator instead. Maybe rewrite the `RedisConfigForm` to use `ipl\Html\Form` instead to avoid making another wrapper.

**Comment by Al2Klimov** [comment]
Does this sound like a plan?

Bildschirmfoto 2021-05-12 um 16 56 07
Bildschirmfoto 2021-05-12 um 16 56 21
Bildschirmfoto 2021-05-12 um 16 56 33

To be honest I consider hiding the insecure option until validation failure not very intuitive.

**Comment by nilmerg** [comment]
Is this now ready?

**Comment by nilmerg** [comment]
Oh, wait, your question... It may not be intuitive, but it shouldn't be seen as suggested solution either way.

It's not possible to set the insecure flag at first, especially not if it's not necessary. Only if it fails to connect due to that very reason, it should be possible to ignore the self-signed certificate.

This prevents the user from setting the flag without caring that it is necessary. It's also more clear that it's not recommended in any way, not actively nor passively.

File: application/forms/RedisConfigForm.php [hunk 1]
```diff
@@ -169,16 +298,22 @@ public function addSubmitButton()
 
     public static function checkRedis($form)
     {
-        $streamy = version_compare(phpversion('redis'), '4.3.0', '>=');
```
Comment (nilmerg) [inline application/forms/RedisConfigForm.php:172]: Why did you remove that and everything related to it? What does this have to do in any way with TLS support?

Restore this. It's purpose is that Icinga DB Web still functions *kind of* on PHP 5.6 installations.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: application/forms/RedisConfigForm.php (hunk 1 above)
Comment (Al2Klimov) [reply application/forms/RedisConfigForm.php:172]: But this PR drops PHP-Redis usage?

Comment (nilmerg) [reply application/forms/RedisConfigForm.php:172]: Aye.. didn't notice that. Would have been good if you mentioned this in the OP. It's an architectural change after all.

File: application/forms/RedisConfigForm.php [hunk 2]
```diff
@@ -137,6 +245,27 @@ public function isValidPartial(array $formData)
...
+            $key = $this->getElement('redis_key');
+
+            if (($cert !== null && $cert->getValue() !== '') !== ($key !== null && $key->getValue() !== '')) {
+                $this->addError(t(
+                    'Either both a client certificate and its private key or none of them must be specified'
+                ));
+
+                return false;
+            }
+        }
```
Comment (nilmerg) [inline application/forms/RedisConfigForm.php:264]: Do that in method `isValid()`.

File: application/forms/RedisConfigForm.php [hunk 3]
```diff
@@ -18,11 +26,79 @@ public function init()
...
+                'description' => sprintf(
+                    t('Authenticate using this PEM-encoded private key ("%s...")'),
+                    '-----BEGIN PRIVATE KEY-----'
+                ),
+                'validators'  => [$this->wrapIplValidator(PrivateKeyValidator::class, 'redis_key')]
+            ]);
+        } else {
+            $this->addElement('hidden', 'redis_ca');
+            $this->addElement('hidden', 'redis_cert');
+            $this->addElement('hidden', 'redis_key');
```
Comment (nilmerg) [inline application/forms/RedisConfigForm.php:68]: Why? It works fine for me without them and the error about the required CA is gone as well when enabling TLS.

File: application/forms/RedisConfigForm.php [hunk 4]
```diff
@@ -18,11 +26,79 @@ public function init()
...
+                    'FormElements',
+                    ['HtmlTag', ['tag' => 'div']],
+                    [
+                        'Description',
+                        ['tag' => 'span', 'class' => 'description', 'placement' => 'prepend']
+                    ],
+                    'Fieldset'
+                ],
+                'description' => t(
+                    'Redis connection details of your Icinga host. If you are running a high'
```
Comment (nilmerg) [inline application/forms/RedisConfigForm.php:85]: These are not connection details. Not in the same manner as the other options are and they use this description already. Use `Secure connections. If ...` instead.

**Review by nilmerg — CHANGES_REQUESTED** [review]
* There's a new usage of `IcingaRedis::getLastIcingaHeartbeat()` in class `RedisHealth`. Adjust the return value check there as well.
* There's absolutely no reason to introduce static methods in the `IcingaRedis` trait. It's **always** tied to and used by an object.
* I see a point in the need to use a non-persistent `Config` for the `IcingaRedis` trait. Though, please don't let one pass that to each method separately. Introduce a setter for the configuration instead.

File: library/Icingadb/Common/IcingaRedis.php [hunk 5]
```diff
@@ -69,60 +65,118 @@ public function getIcingaRedis()
         return $this->redis;
     }
 
-    public function getLastIcingaHeartbeat(Redis $redis)
+    public static function getLastIcingaHeartbeat(Redis $redis)
     {
-        if (! self::streamSupportAvailable()) {
-            return false;
-        }
+        // https://github.com/predis/predis/issues/607#event-3640855190
```
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply library/Icingadb/Common/IcingaRedis.php:70]: What is a _passive-aggressive link_?

[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Icingadb/Common/IcingaRedis.php:70]: A link to an issue that has no description and to a close event by the project's author. To me, the link's only purpose is to highlight the author's missing sign of responsibility and motivation.

[the remark this answers is not in the document]
Comment (Al2Klimov) [reply library/Icingadb/Common/IcingaRedis.php:70]: Ah, I see. You've completely misunderstood my intention. (No reason to ascribe anything to me.) My intention was to reason the executeRaw() usage. But now as you mentioned it: the specific issue event indeed doesn’t matter as the issue could also be open, but w/o any fix, yet. I'll shorten the link.

**Comment by Al2Klimov** [comment]
1. Done.
2. Except the case where `RedisPage#isValid()` calls `RedisConfigForm::checkRedis()` (has to be static therefore) which calls `IcingaRedis::*()` (have to be static therefore).
3. See (2).

File: library/Icingadb/Common/IcingaRedis.php [hunk 6]
```diff
@@ -69,60 +65,118 @@ public function getIcingaRedis()
...
+        }
+
+        return ['scheme' => 'tls', 'ssl' => $ssl];
+    }
+
+    private static function ensurePemOnDisk($pem)
     {
-        $extRedisVersion = phpversion('redis');
-        return version_compare($extRedisVersion, '4.3.0', '>=');
+        $dir = sys_get_temp_dir() . DIRECTORY_SEPARATOR . sha1($pem);
```
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply library/Icingadb/Common/IcingaRedis.php:172]: Are you sure the certs shall be written and deleted every single time?

File: application/forms/RedisConfigForm.php (hunk 3 above)
Comment (Al2Klimov) [reply application/forms/RedisConfigForm.php:68]: Haven’t you taught me to do exactly like this? (MSAD vs. OpenLDAP) So I could e.g. switch TLS on, enter certs, switch TLS off, switch TLS on and the certs are still there.

**Comment by nilmerg** [comment]
> Except the case where RedisPage#isValid() calls RedisConfigForm::checkRedis() (has to be static therefore) which calls IcingaRedis::*() (have to be static therefore).

What speaks against a `new self()` inside checkRedis? It's the wizard page why checkRedis needs to be static, not the IcingaRedis trait. So no need to introduce this into the trait as well.

**Comment by Al2Klimov** [comment]
> * I see a point in the need to use a non-persistent `Config` for the `IcingaRedis` trait. Though, please don't let one pass that to each method separately. Introduce a setter for the configuration instead.

I've reduced unnecessary `Config` params. Are you sure it’s a good idea for independent method calls to possibly interfer?

File: library/Icingadb/Common/IcingaRedis.php [hunk 7]
```diff
@@ -6,13 +6,61 @@
...
+        }
+
+        if ($redis2 !== null && $this->getLastIcingaHeartbeat($redis2) === null) {
+            $this->warning(t('Secondary connection established but failed to verify Icinga is connected as well.'));
+            return false;
+        }
+
+        $this->info(t('The configuration has been successfully validated.'));
+        return true;
+    }
```
Comment (nilmerg) [inline library/Icingadb/Common/IcingaRedis.php:62]: This was not part of the trait, why is it one now? This doesn't even make any sense, since this method can only be utilized if the trait is being used by a form. This implicit dependency doesn't suit in a trait which's usecase is not only for forms.

Remove that and keep it where it was.

File: library/Icingadb/Common/IcingaRedis.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Icingadb/Common/IcingaRedis.php:218]: No, if you are talking about every single redis request. Yes, if we're talking about every user request.

File: application/forms/RedisConfigForm.php (hunk 3 above)
Comment (nilmerg) [reply application/forms/RedisConfigForm.php:68]: I see.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Icingadb/Common/IcingaRedis.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Icingadb/Common/IcingaRedis.php:218]: Well, it's not just a performance question. The reason why I don't see an issue here, is that we're putting private keys (secrets) onto disk there. But that's also the case if they're part of the ini configuration... This can only be accessed by the owner and group, and so it's the case for the storage dir... (2770 in package installations)

Well, yes, a regular file storage sounds okay? What's your proposal regarding the actual path and structure?

[the remark this answers is not in the document]
Comment (Al2Klimov) [reply library/Icingadb/Common/IcingaRedis.php:218]: Ah, security is the problem! What if I just umask()ify my current implementation?

[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Icingadb/Common/IcingaRedis.php:218]: No. We are on a nice track with the storage dir. A perfect chance to use that with a proper use-case and implementation.

[the remark this answers is not in the document]
Comment (Al2Klimov) [reply library/Icingadb/Common/IcingaRedis.php:218]: I'd suggest the framework to allow modules `$STORAGE/modules/$MODULE` and this module to use `$STORAGE/modules/$MODULE/redis/{ca,cert,key}.pem` for saved config and temp files for not yet saved one.

[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Icingadb/Common/IcingaRedis.php:170]: Sounds fine to me.

**Review by nilmerg — APPROVED** [review]
